### ISSUE TITLE
`sample_weight` crashes on `train_on_batch`

### ISSUE THREAD

**Issue by EderSantana** [fchollet/keras#1199]
Reproduce error with
```python
model = Sequential()
model.add(LSTM(128, return_sequences=True, input_dim=128))  # try using a GRU instead, for fun
model.compile('sgd', 'mse')

X = np.random.randn(100, 10, 128)
model.train_on_batch(X, X, sample_weight=np.ones_like(X))
```

We used to be able to weight batch and time dimensions. Now we can only weight batch dim? How do we consider which points in time should go out? This used to be key to the Neural Turing Machines.

It does not crash if use fit:
```
model.fit(X, X, sample_weight=np.ones_like(X))
```

`train_on_batch` and `fit` used to have similar APIs. What changed? I can write the tests and possible fixes if I get some guidance.

**Comment by fchollet** [comment]
That's not supposed to work, sample weights are for samples (as the name may indicate), not for timesteps. 

The previous implementation touched many functions, was counter-intuitive and complex, which is why it has been removed. If we reintroduce timestep weighting, it would have to be simple and straightforward.

In any case, it should not crash, and the invalid use of the argument should throw an exception. So that should be fixed.

**Comment by EderSantana** [comment]
I understand. Is there any way the user can control time weighting? only with masking right?

**Comment by fchollet** [comment]
Yes. One possibility would be to make masking continuous, but it's not clear if that would make much sense given that existing masking layers are binary.

**Comment by EderSantana** [comment]
Its ok, I only need binary mask anyway. I've honestly just learned to not rely on masks. But let me give the new API a try :)

Thanks!

**Comment by EderSantana** [comment]
But one thing, sample_weight is expected to be of the same shape of `y_train`, right? What happens to the other values them, if only samples dim is considered?

I'll read the code more carefully. Thanks again!

**Comment by fchollet** [comment]
I've noticed that the current implementation does in fact allow for a continuous mask. But not that the total loss is reweighted by the mean of the mask, which doesn't happen for sample weighting (but, thinking about it, maybe it should also be reweighted in the latter case). Anyway, that's a small detail.

Anyway, better error messages are needed for the sample weighting interface.
